controllers/user/login.js

```diff
@@ -44,33 +58,63 @@ router
         });
         renderForm(req, res);
     })
-    .post((req, res, next) => {
+    .post(async (req, res, next) => {
         logger.info('Login attempted');
-        passport.authenticate('local', function(err, user) {
-            if (err) {
-                logger.error('Authentication failed', err);
-                next(err);
-            } else if (user) {
-                req.logIn(user, function(loginErr) {
-                    if (loginErr) {
-                        logger.error('Login failed', loginErr);
-                        next(loginErr);
-                    } else {
-                        logger.info('Login succeeded');
-                        redirectUrlWithFallback(req, res, '/user');
+
+        // Key all rate-limited login attempts by the user's IP and the attempted username
+        const uniqueKey = `${req.ip}_${req.body.username}`;
+
+        const LoginRateLimiter = await new RateLimiter(
+            availableRateLimiters.failByUsername,
+            uniqueKey
+        ).init();
+
+        if (LoginRateLimiter.isRateLimited()) {
+            // User is rate limited
+            return renderRateLimitError(req, res);
+        } else {
+            passport.authenticate('local', async function(err, user) {
+                if (err) {
+                    logger.error('Authentication failed', err);
+                    next(err);
+                } else if (user) {
+                    req.logIn(user, async function(loginErr) {
+                        if (loginErr) {
+                            logger.error('Login failed', loginErr);
+                            next(loginErr);
+                        } else {
+                            logger.info('Login succeeded');
+                            // Reset rate limit on successful authorisation
+                            if (LoginRateLimiter.hasConsumedPoints()) {
+                                await LoginRateLimiter.clearRateLimit();
+                            }
+                            redirectUrlWithFallback(req, res, '/user');
+                        }
+                    });
+                } else {
+                    /**
+                     * User is invalid
+                     * Show a generic error message here to avoid exposing account state
+                     */
+                    logger.warn('Login failed: invalid credentials');
```

Comment:
Just realised we have `Login failed` as the log value both for this case and also when the login fails due to an error. Could we renamed this to something like `Unsuccessful login attempt` to help disambiguate these two.